chapter_attention-mechanisms-and-transformers/attention-pooling.md

```diff
@@ -54,349 +38,152 @@ from d2l import torch as d2l
-:eqref:`eq_nadaraya-watson` gives
-
-$$\begin{aligned} f(x) &=\sum_{i=1}^n \alpha(x, x_i) y_i\\ &= \sum_{i=1}^n \frac{\exp\left(-\frac{1}{2}(x - x_i)^2\right)}{\sum_{j=1}^n \exp\left(-\frac{1}{2}(x - x_j)^2\right)} y_i \\&= \sum_{i=1}^n \mathrm{softmax}\left(-\frac{1}{2}(x - x_i)^2\right) y_i. \end{aligned}$$
-:eqlabel:`eq_nadaraya-watson-gaussian`
-
-In :eqref:`eq_nadaraya-watson-gaussian`,
-a key $x_i$ that is closer to the given query $x$ will get
-*more attention* via a *larger attention weight* 
-assigned to the key's corresponding value $y_i$.
-
-The Nadaraya-Watson kernel regression is a nonparametric model;
-thus :eqref:`eq_nadaraya-watson-gaussian`
-is an example of *nonparametric attention pooling*.
-In the following, we plot the prediction based on this
-nonparametric attention model.
+Now that we have data and kernels, all we need is a function that computes the kernel regression estimates. Note that we also want to obtain the relative kernel weights, aka, the attention weights, in order to perform some minor diagnostics. Hence we first compute the kernel between all training covariates `x_train` and all test covariates `x_test`. This yields a matrix, which we subsequently normalize. When multiplied with the training labels `y_train` we obtain the estimates.
 
 ```{.python .input}
 %%tab all
-def diff(queries, keys):
-    return d2l.reshape(queries, (-1, 1)) - d2l.reshape(keys, (1, -1))
-
-def attention_pool(query_key_diffs, values):
-    if tab.selected('mxnet'):
-        attention_weights = npx.softmax(- query_key_diffs**2 / 2, axis=1)
+def watson_nadaraya(x_train, y_train, x_test, kernel):
+    dists = d2l.reshape(x_train, (-1, 1)) - d2l.reshape(x_test, (1, -1))
+    k = kernel(dists) # compute attention mask
+    k = k / k.sum(0)  # normalization
     if tab.selected('pytorch'):
-        attention_weights = F.softmax(- query_key_diffs**2 / 2, dim=1)
+        y_hat = y_train@k
+    if tab.selected('mxnet'):
+        y_hat = np.dot(y_train, k)
     if tab.selected('tensorflow'):
-        attention_weights = tf.nn.softmax(- query_key_diffs**2/2, axis=1)
-    return d2l.matmul(attention_weights, values), attention_weights
-
-y_hat, attention_weights = attention_pool(
-    diff(data.x_val, data.x_train), data.y_train)
-plot_kernel_reg(y_hat)
+        y_hat = tf.matmul(y_train, k)
+    return y_hat, k
 ```
 
-Now let's take a look at the [**attention weights**].
-Here validation inputs are queries while training inputs are keys.
-Since both inputs are sorted,
-we can see that the closer the query-key pair is,
-the higher attention weight is in the attention pooling.
+Let's have a look at the kind of estimates that the different kernels produce.
 
 ```{.python .input}
-%%tab all
-d2l.show_heatmaps([[attention_weights]],
-                  xlabel='Sorted training inputs',
-                  ylabel='Sorted validation inputs')
+def plot_results(x_train, y_train, x_test, y_test, kernels, names, attention = False):
```

Comment:
I think we need to add back other frameworks as well. MxNet and Tensorflow are removed and are broken in the current state.